### PR TITLE
Condense parsing code by using functional constructs and if/when as expression

### PR DESCRIPTION
**Pull request by MikkCZ**
Parser was the last part of the codebase I was not familiar with (probably the most complex), and this is the last PR of this type. After that, I a think I am ready to implement something new too.

### CODE REVIEW

File: fluent.syntax/src/main/kotlin/org/projectfluent/syntax/parser/Stream.kt [hunk 1]
```diff
@@ -52,14 +57,10 @@ internal open class ParserStream(var string: String) {
...
-        this.peekOffset = 0
+        resetPeek()
     }
 }
 
-internal const val EOL = '\n'
-internal val EOF: Char? = null
-internal const val SPECIAL_LINE_START_CHARS = "}.[*"
-
 internal class FluentStream(string: String) : ParserStream(string) {
```
Comment (MikkCZ) [inline fluent.syntax/src/main/kotlin/org/projectfluent/syntax/parser/Stream.kt:64]: Is there actually need for `ParserStream` to be standalone class? I understand it wraps string operation pretty nicely. However the only use is here in the `FluentStream`. May I suggest to use composition over inheritance then?

Comment (MikkCZ) [reply fluent.syntax/src/main/kotlin/org/projectfluent/syntax/parser/Stream.kt:64]: Btw. unit tests exist for `ParserStream` only, and they run with `ParserStream` instance, not the final implementation of `FluentStream`, which can actually override it all. `FluentStream` is only tested as an implementation detail of `Parser`, never as a unit (in `SerializeResourceTest`).

**Comment by MikkCZ** [comment]
Hey @eemeli @Pike , is there still any interest to keep `fluent-kotlin` alive? Or should I close my PRs as something that's not interesting anymore?

**Comment by eemeli** [comment]
@MikkCZ As you've gathered, `fluent-kotlin` has not received much love and care for a while.

If you're interested in helping to maintain it, I'd be happy to give you admin access to the repo?

**Pull request closed without merging**